In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [4]:
df_ords = pd.read_csv(r'/Users/smw/Desktop/9-29-20 Instacart Basket Analysis/02 Data/Original Data/orders.csv', index_col = False)

In [5]:
path = r'/Users/smw/Desktop/9-29-20 Instacart Basket Analysis'

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [7]:
df_prods = pd.read_csv(r'/Users/smw/Desktop/9-29-20 Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [8]:
path = r'/Users/smw/Desktop/9-29-20 Instacart Basket Analysis'

In [9]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
# Create a dataframe

df_test = pd.DataFrame()

In [11]:
# Create a mixed type column
df_test['mix'] = ['a', 'b' , 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [1]:
# Check for Mixed Types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

NameError: name 'df_test' is not defined

In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [19]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [20]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [21]:
df_prods.shape

(49693, 5)

In [22]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
df_prods_clean.shape

(49677, 5)

In [24]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##QUESTION 2##  The max for prices is way up to 99999.000000 
while the mean is only at 9.994136 this seems very unusal to me (I would
think that the mean would be much higher so there must be a mistake with
the max price.)

In [48]:
for col in df_ords.columns.tolist():  ##QUESTION 3 ##
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [27]:
# Check for Mixed Types    ##QUESTION 3#
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [50]:
df_ords.info

<bound method DataFrame.info of          order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12              

In [30]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##QUESTION 5## The first time shopping for every customer has a null "days_since_prior_order"
are there are also customers that only shop a single time which is included in that.

In [29]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [31]:
df_nan_2 = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [32]:
df_nan_2

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [32]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

##QUESTION 6##. I decided to change all of the Nan's to "0" since it merely means
that it is the customer's first order and that means there has been 0 days since
the last order.

In [36]:
df_ords.shape

(3421083, 7)

In [40]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [41]:
df_ords_clean.info

<bound method DataFrame.info of          order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12              

In [42]:
df_ords_clean.shape

(3421083, 7)

In [43]:
df_dups_2 = df_ords_clean[df_ords_clean.duplicated()]

In [44]:
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [46]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.044488e+01  
std         4.226088e+00            9.308727e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [47]:
df_dups_2

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

##QUESTIONS 7 and 8## No dupicates found.

In [51]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [52]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))